slapo/pipeline.py

```diff
@@ -274,6 +275,78 @@ def get_itemized_name(node, suffix=""):
     return stage_id_2_arg_names, stage_id_2_name, liveness
 
 
+def analyze_tie_weights(top_mod):
+    """Analyze if there is any tie weights (two weights in different module
+    share the same memory) partitioned into different pipeline stages.
+
+    Parameters
+    ----------
+    top_mod : torch.nn.Module
+        The top-level module. This should be a top pipeline module, so
+        1) it should already be traced and partitioned, and
+        2) it should have a number of submodules that matches pipeline stages.
+
+    Returns
+    -------
+    tie_groups : Dict[str, Set[Tuple[str, int]]]
+    """
+    # Mapping from parameter name to (the pipeline stage ID, the parameter object ID).
+    params = {}
+    # Mapping from the primary key (i.e., the first parameter name)
+    # of a tie group to the tie group.
+    tie_groups = {}
+    # Mapping from paramter name to the primary key of the tie group.
+    param_name_2_group_key = {}
+
+    def _traverse_children(stage_id, prefix, curr_mod_name, curr_mod):
+        full_prefix = f"{prefix}.{curr_mod_name}" if prefix else curr_mod_name
+
+        for curr_param_name, curr_param in curr_mod.named_parameters():
+            curr_param_full_name = f"{full_prefix}.{curr_param_name}"
+            curr_param_id = id(curr_param)
+            if curr_param_full_name in params:
+                continue
+
+            # Check if this parameter is tie to another one in a different stage.
+            for target_param_full_name, (
+                target_stage,
+                target_param_id,
+            ) in params.items():
+                if stage_id != target_stage and curr_param_id == target_param_id:
+                    if target_param_full_name in param_name_2_group_key:
+                        # Get the tie group of the target parameter.
+                        tie_group_key = param_name_2_group_key[target_param_full_name]
+                        tie_group = tie_groups[tie_group_key]
+                    else:
+                        # Create a new tie group, and use the target parameter name
+                        # as the primary key.
+                        param_name_2_group_key[
+                            target_param_full_name
+                        ] = target_param_full_name
+                        tie_group = set([(target_param_full_name, target_stage)])
+                        tie_groups[target_param_full_name] = tie_group
+
+                    # Add the current parameter to the tie group.
+                    param_name_2_group_key[
+                        curr_param_full_name
+                    ] = target_param_full_name
+                    tie_group.add((curr_param_full_name, stage_id))
```

Comment:
Set is missing

Reply:
What set?

Reply:
tie_group value is a Set based on your docstring and the other part of the code.

Reply:
So tie_group.add adds a new value to the set. I didn't see the problem?

Reply:
Oh nvm, I look at the wrong variable.